lib/timeline/timeline.go

```diff
@@ -0,0 +1,280 @@
+/*
+Copyright 2019 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package timeline provides interfaces for keeping track of cluster status events.
+package timeline
+
+import (
+	"fmt"
+	"strings"
+	"time"
+
+	pb "github.com/gravitational/satellite/agent/proto/agentpb"
+)
+
+// Stamp defines default timestamp format for events.
+const Stamp = "Jan _2 15:04:05"
+
+// Event represents a Timeline event. An event occurrs  whenever there is a
+// change in the cluster status. An event could be triggered by a failed
+// heartbeat or a failed health check.
+type Event struct {
+	// TimeStamp specifies when the event occurred.
+	TimeStamp time.Time
+
+	// TODO: can we rely on user CLI sessions to provide colored text?
+	// If not, maybe we should indicate the different event types a different
+	// way.
+
+	// Color specifies the color of the event.
+	// Red -> Indicates change has caused state to degraded.
+	// Yellow -> Indicates change has not caused any state changes.
+	// Green -> Indicates changes has caused state to recover.
+	Color string
+
+	// Description specifies a description of the event.
+	Description string
+}
+
+// NewEvent initializes and returns a new Event with the current timestamp.
+func NewEvent() Event {
+	return Event{
+		TimeStamp: time.Now(),
+	}
+}
+
+// String returns a string representation of Event.
+func (e *Event) String() string {
+	var sb strings.Builder
+	sb.WriteString(fmt.Sprintf("[%s] ", e.TimeStamp.Format(Stamp)))
+	sb.WriteString(e.Description)
+	return sb.String()
+}
+
+// Timeline represents a timeline of cluster status events. The Timeline
+// can hold a specified amount of events and uses a FIFO eviction policy.
+type Timeline struct {
+	// Size specifies the max size of the timeline.
+	Size int
+	// Timeline holds the latest status events.
+	Timeline []string
```

Comment:
I think this should also be something structured instead of just strings. I.e. I would define an `Event` gRPC object that would be returned in timeline via gRPC API and use that here as well.